### ISSUE TITLE
[Feature] Slow sources benchmark and timeout

### ISSUE THREAD

**Issue by ray-x** [hrsh7th/nvim-cmp#148]
As I adding more sources to cmp the cmp is getting slower.
Is it possible poll the source in parallel and timeout the slow one? Also would be good to show some statistics on the source performance.  I have no clue which source is slow and should be turned off. (tablenine is an obvious one)

**Comment by hrsh7th** [comment]
Benchmark is possible to implement.
But parallel request is impossible in current neovim.

**Comment by Pocco81** [comment]
@ray-x I'm experiencing the same, but your issue gave me an idea. Since according to @hrsh7th parallel requests are not possible, then what about lazy loading the completion elements?

For example: [cmp-spell](https://github.com/f3fora/cmp-spell) gives spell suggestions using vim's `spellsuggest`, which ends up giving a vast amount of completion items and hence slows down nvim (mostly when typing). @hrsh7th would it be possible to only "pull", for instance, the first 5 completion items from _x_ source, and then when nvim-cmp's popup/floating window is scrolled load another five and so on and so forth?

**Comment by Pocco81** [comment]
> I have no clue which source is slow and should be turned off

In my case [cmp-spell](https://github.com/f3fora/cmp-spell) and [cmp-path](https://github.com/hrsh7th/cmp-path).

**Comment by hrsh7th** [comment]
I don't think spellsuggest returns a lot of items. How to reproduce such behavior?

**Comment by Shougo** [comment]
Vim spellsuggest API is slow and it blocks your neovim.

Path source is depends on disk speed.

**Comment by Pocco81** [comment]
Basically what @Shougo said. To reproduce the spellsuggest behaviour, well, add the `cmp-spell` source and just type fast or keep pressing backspace in a markdown file, you should be able to see some "lag".

**Comment by hrsh7th** [comment]
If spellsuggest itself is slow, there's nothing we can do because spellsuggest API is vim native and only allows synchronous call, right?

The path source is using Luv API so it work as parallel (partially).

**Comment by hrsh7th** [comment]
The `slow` is not the same meaning as `lag`.


> Path source is depends on disk speed.

This means `slow`. Because cmp-path uses `Luv` API so it wont cause the lag.

> Vim spellsuggest API is slow and it blocks your neovim.

This can be the reason to the lag.

**Comment by ray-x** [comment]
Some suggestions:
1) When enforce timeout, set the priority for sources based on sources array index when timeout is used.  e.g.
```lua
sources = {
      {name = 'nvim_lsp'}, {name = 'buffer'},  {name = 'look'},
    }
```
That will let cmp get items from lsp first and then buf and then look.
When timeout,  drop the result if low priority sources do not finish.

Alternatively use the `priority` field.

2) disable/enable sources for filetype type. Instead of using autocmd on FileType suggested in README.md use this:

e.g.
```
sources = {
      {name = 'nvim_lsp'}, {name = 'buffer'}, {name = 'nvim_lua', ft='lua'},{name = 'look', ft={'markup','html'}},
    }
```
this will only enable `nvim_lua` for lua, and `look` for markdown and html file types.
This setup will be more consistent with other setups for sources, e.g. `keywork_length` setup.

**Comment by hungrybirder** [comment]
The same issue.
I found that `nvim-cmp` is a little bit slower than `nvim-compe`

**Comment by hrsh7th** [comment]
I can't figure out it. How slow it?

**Comment by Pocco81** [comment]
> I found that nvim-cmp is a little bit slower than nvim-compe

Isn't that relative to the amount of sources/which sources you are using?

**Comment by hungrybirder** [comment]
> > I found that nvim-cmp is a little bit slower than nvim-compe
> 
> Isn't that relative to the amount of sources/which sources you are using?

https://github.com/hungrybirder/dotfiles/blob/using_nvim_cmp_but_it_is_slower/nvim/lua/module/nvim-cmp.lua

```lua
  sources = {
    { name = 'nvim_lsp'},
    { name = 'path'},
    { name = 'nvim_lua'},
    { name = 'vsnip'},
    { name = 'calc'},
    { name = 'emoji'},
    { name = 'tags'},
  },
```

But I use the same sources in [`nvim-compe`](https://github.com/hungrybirder/dotfiles/blob/master/nvim/lua/module/nvim-compe.lua) and it's really fast.

**Comment by Shougo** [comment]
Can you upload the examples?

**Comment by ray-x** [comment]
I have a feeling the difference is caused by `throttle_time = 80` setup.

**Comment by hungrybirder** [comment]
> I have a feeling the difference is caused by `throttle_time = 80` setup.

Is there a way to setup `throttle_time` in `nvim-cmp`?

**Comment by hrsh7th** [comment]
I've improved the performance. Could you test with latest?

And throttle_time was removed. Currently, I'd specified 50ms as a hardcoded value.

**Comment by ray-x** [comment]
I suspect the following codes:
https://github.com/hrsh7th/cmp-buffer/blob/5dde5430757696be4169ad409210cf5088554ed6/lua/cmp_buffer/buffer.lua#L43
https://github.com/hrsh7th/cmp-buffer/blob/5dde5430757696be4169ad409210cf5088554ed6/lua/cmp_buffer/init.lua#L59

need to be double-checked.
From what I see. If the buffer source is in processing state, the callback will be updated after `50ms`(hardocded). 100ms timer and 200ms timer has some impact on the overall performance.

**Comment by hungrybirder** [comment]
> I've improved the performance. Could you test with latest?
> 
> And throttle_time was removed. Currently, I'd specified 50ms as a hardcoded value.

I use commit b6b15d5f6e46643462b5e62269e7babdab17331c.
It works well.
Now I switch `nvim-compe` to `nvim-cmp`.
😄

**Comment by Pocco81** [comment]
@hungrybirder since you just migrated from compe, I'd recommend ya my [cmp + luasnip + friendly snippets + autopairs setup](https://github.com/Pocco81/TheSupercalifragilisticexpialidociousDots/blob/main/.config/nvim/lua/nvdope/initialization/init.lua#L196-241) and my [cmp config](https://github.com/Pocco81/TheSupercalifragilisticexpialidociousDots/blob/main/.config/nvim/lua/nvdope/initialization/tools/cmp.lua) ;)

**Comment by hungrybirder** [comment]
> @hungrybirder since you just migrated from compe, I'd recommend you my [cmp + luasnip + friendly snippets + autopairs setup](https://github.com/Pocco81/TheSupercalifragilisticexpialidociousDots/blob/main/.config/nvim/lua/nvdope/initialization/init.lua#L196-241) and my [cmp config](https://github.com/Pocco81/TheSupercalifragilisticexpialidociousDots/blob/main/.config/nvim/lua/nvdope/initialization/tools/cmp.lua) :)

I'm using cmp + autopairs + (cmp-vsnip+vsnip+friendly-snippets).
I'll check luasnip later and learn your config. 

Thank you.

**Comment by nyngwang** [comment]
Is it possible to cache all the completion from sub-packages from the start? E.g. I'm writing Python with TensorFlow package, and after I entered `tf.keras.` I need to wait for about 1~2s for the completion menu to show up.

**Comment by Shougo** [comment]
It seems LSP slowness.
pylsp use jedi and it is slow.

You can use pyright lsp instead.

**Comment by nyngwang** [comment]
@Shougo: It seems that pyright will cause high-cpu usage so I change to use pylsp. Anyway, I will try out pyright again.

**Comment by Shougo** [comment]
> Is it possible to cache all the completion items from sub-packages from the start? 

I think pyright implements this feature, and it uses CPU usage.
So it is trade off.

**Comment by nyngwang** [comment]
> [...]
> 
> You can use pyright lsp instead.

@Shougo: I just confirmed that pyright is indeed (much) faster. Thanks for your help!

**Comment by nyngwang** [comment]
> [...] Also would be good to show some statistics on the source performance [...]
> 
> [(scroll to top)](https://github.com/hrsh7th/nvim-cmp/issues/148#issue-988198126)

While this might be off-topic, but it would be cool if these statistics can be shown on the status line, or even be combined with plugin like nvim_notify.
